### ISSUE TITLE
SyntaxError: invalid syntax

### ISSUE THREAD

**Issue by bane77111** [JonasSchatz/DepixHMM#1]
Hi

When i try to install requirements i get SyntaxError: invalid syntax,... Any solution for this? Thsnks
<img width="211" alt="Screen Shot 2021-10-12 at 18 33 38" src="https://user-images.githubusercontent.com/81524585/136995530-88a2d505-8534-4a41-8429-8906f187a2fd.png">

**Comment by JonasSchatz** [comment]
I've added a bit more detailed installation instructions to the readme. 

If this does not help, reach out with more information of what you tried and what exactly went wrong. The error message (SyntaxError being a python-thing) indicates to me that you try installation through python? You should type the installation commands into your (non-python) terminal.

**Comment by bane77111** [comment]
> I've added a bit more detailed installation instructions to the readme.
> 
> If this does not help, reach out with more information of what you tried and what exactly went wrong. The error message (SyntaxError being a python-thing) indicates to me that you try installation through python? You should type the installation commands into your (non-python) terminal.

Hi againNow i have all requirements, but cos i m not computer/IT guy, i have no clue how to run this script on terminal. To make a long story short; my not so nice brother-get my pass from account,... change password and send me this (attachment) (just for fun???) i get it beck when he returns (in three weeks) so is there a chance to see the password?
Thanks a lot in advance, best 
Seb
<img width="239" alt="Screen Shot 2021-10-18 at 09 34 01" src="https://user-images.githubusercontent.com/81524585/137687831-24321a7f-064f-4399-9c86-e31eeffa8fb2.png">

**Comment by JonasSchatz** [comment]
It's been three weeks since you opened the issue, I hope you got your password back (even though the story sounds a bit sketchy, to be honest). I think decoding the image you have sent me would be possible, however it would require more time than I currently have:
- I would need to figure out the exact font parameters (which font, which size, but that doesn't have to 100% accurate)
- Some minor code changes would be needed: So far the tool only works with black text on white background, but making those two variables isn't an issue
- Then, and this is the time-comsuming part: Experiment to find the right training parameters to find the underlying text

Generally speaking, this is the right tool for the job (the original [Depix](https://github.com/beurtschipper/Depix) library might also get it done, and has the capabilities to work with arbitrary colors right out of the box), and I'm REALLY curious to try it, since I appreciate the challenge. Don't expect it in the next few days, though.

**Comment by bane77111** [comment]
Hi Jonas

Thanks for your reply and willing to help. I try with Depix but didnt get
any resolt,...  You dont need to bother i get my pass beck, but if you like
the chalange you can try it,...
ps. yes sketchy family🙂

Thanks again and all the best,
Seb

**Comment by JonasSchatz** [comment]
I managed to play around with your image a bit more. 
TL;DR: I didn't manage to decode it. There were issues with the image itself, and there might be some more with my code. If someone encounters this in the future, I'll describe it below.

## Issues with the original image

1. Unknown Font
In order to train the HMM, we need to generate data with the same font type and size as the pixelized text. I had very little to work with (only the first `a`). I decided to go with Arial, size 30. It's not a perfect fit, as you can see, but it was surprisingly hard to find a better one (btw, while this is slightly too wide, Arial _narrow_ was, well, too narrow):
![grafik](https://user-images.githubusercontent.com/13544864/146993011-71b351b9-f928-471b-84d3-0f027adbe416.png)

2. Compression / Resizing Issues:
When zooming in, you can see that there are no sharp edges. This might be due to resizing, as the transitions between the regions is pretty smooth:
![grafik](https://user-images.githubusercontent.com/13544864/146993372-a84ec7e9-5025-436e-a63b-f84838d2fc19.png)

I then created a "remastered" version by applying a pixelization filter:
![grafik](https://user-images.githubusercontent.com/13544864/146993518-1fc7ffd2-623d-4c5b-b9d2-ba978abfd8ca.png)

This brings us to the third point.

3. Non-integer pixlization size
I measured the blocks in the original image and found that they have a size of almost exactly 9.5 px, averaging over the whole thing. This is an issue, probably  caused by the resizing. You could just scale up the whole thing by a factor of 2 and use a block size of 19, but I noticed too late. 

## Issues with the Code
First of all, I had to make the background and font color a parameter, but that was no problem. I then let it run over night, but didn't get any sensible information. According to the original paper, we are still in a range of the ratio font size / block size where we would expect a reasonable result. It is possible that I just used wrong parameters for training. It is also possible that there is something slightly wrong with the code, especially where we assign the unknown slices to a cluster. 

## Summary
I'll close the issue for now. Thanks for providing me with this example, I had fun testing it. This little image will probably remain my nemesis for a while, but it's good to have a challenge and some corpus to test with. 
The samle code for this will be under `experiments`.
